### ISSUE TITLE
provide docker based installation

### ISSUE THREAD

**Issue by carueda** [stoqs/stoqs#552]
It would be good to provide a docker-based installation of the STOQS system.

Relate: [this comment in issue 57](https://github.com/stoqs/stoqs/issues/57#issuecomment-79794124)

**Comment by MBARIMike** [comment]
Agreed!

Perhaps we can start with a Dockerfile added to the python3 branch (tracked on Issue https://github.com/stoqs/stoqs/issues/523)?

**Comment by carueda** [comment]
Thanks Mike.

As discussed offline, I started a separate `stoqs-docker` repo for this, which has the new docker related stuff, while having `stoqs/stoqs` as a _submodule_.  

I like the submodule approach--as opposed to adding the docker stuff to the main repo-- because the dockerized version is, at least to some extent, a separate facet for the project. The separate space is good at least initially while a working set of images can be put in place.  I'd be happy to make my `stoqs-docker` repo public under my `github.com/carueda/` space or under `bitbucket.org/mbari/` or (perhaps preferably) under  `github.com/stoqs/`.  Please let me know how you would like to proceed.

**Comment by carueda** [comment]
Of course, if you prefer the docker related files to be under this repo (perhaps under a `docker/` subdirectory), that would be just fine too.

**Comment by MBARIMike** [comment]
I think it'd be nice to keep it all together in this repo, perhaps in a `docker/`subdirectory. We can also move to the `provision.sh` and `Vagrantfile` files to a `vagrant/` subdirectory.

**Comment by carueda** [comment]
OK, PR https://github.com/stoqs/stoqs/pull/555

**Comment by MBARIMike** [comment]
A reference for this issue: http://www.eidel.io/2017/07/10/dockerizing-django-uwsgi-postgres/

**Comment by carueda** [comment]
Hi @MBARIMike @duane-edgington , 

Per our meeting a few days ago, I'll be looking into https://hub.docker.com/r/camptocamp/postgis/ (actually, https://hub.docker.com/r/camptocamp/postgres/, as the other one is deprecated).

I think we said this would possibly be a strong basis for key STOQS components.

I shall report here findings about at least some basic testing with camptocamp/postgres, but any comments in the mean time are welcome.

**Comment by MBARIMike** [comment]
Thanks @carueda !

I'm doing development now on the python3 branch where I'm [using PostgreSQL 9.6 and PostGIS 2.3](https://github.com/stoqs/stoqs/blob/f700e8b6b52d99ceab16ceefac279975adde93ac/provision.sh#L247-L260).

It appears that the camptocamp image uses PostgreSQL 9.5 and PostGIS 2.3: https://hub.docker.com/r/camptocamp/postgres/~/dockerfile/

Am I reading this correctly?

**Comment by carueda** [comment]
Yes, that one seems based on postgres:9.5.

I found this: https://hub.docker.com/r/mdillon/postgis/
It has 198 stars, +1M pulls, and the github repo has 194 stars, but more interestingly it seems to have an image based on postgres 9.6 and postgis 2.3, https://github.com/appropriate/docker-postgis/blob/master/9.6-2.3/Dockerfile . So, I guess I'll look into this one a bit closer.

**Comment by carueda** [comment]
@MBARIMike Mike, your input is needed to properly define the necessary images according to interdependencies between the various services.

**Comment by carueda** [comment]
Service definitions still to be completed (and not interlinked yet), but at least to start having a sense of where this is going:

The images currently look like this:

```shell
$ docker images
REPOSITORY              TAG                  IMAGE ID            CREATED             SIZE
mbari/stoqs             0.0.1                f46f2d1bdc52        About an hour ago   13.5GB
mbari/stoqs-mapserver   0.0.1                d3300314ca7f        12 hours ago        11.3GB
mbari/stoqs-base        0.0.1                7adf86acb02b        13 hours ago        11.3GB
mbari/stoqs-postgis     0.0.1                5c143d25ee30        16 hours ago        610MB
rabbitmq                3                    8134ff2de910        8 days ago          124MB
centos                  7                    196e0ce0c9fb        13 days ago         197MB
mdillon/postgis         9.6                  c5ca45c83515        8 weeks ago         610MB
```

Running the system (using Docker Compose):

```shell
$ docker-compose up -d
Creating network "docker_default" with the default driver
Creating stoqs-postgis ...
Creating stoqs-rabbitmq ...
Creating stoqs-postgis
Creating stoqs-postgis ... done
Creating stoqs ...
Creating stoqs ... done
```

The STOQS container includes Apache and MapServer. With the current 80:80 port mapping on my local machine, I can open and see the following:
- `http://localhost` - the usual default Apache "Testing 123.." page
- `http://localhost/cgi-bin/marserv` - "No query information to decode. QUERY_STRING is set, but empty."

**Comment by MBARIMike** [comment]
The Apache httpd and mapserv responses look good.

If targeting for production installation I think we should also have a stoqs-nginx image. I think this can be a standard nginx image that provides a uWSGI "gateway" to the STOQS application. Here is the [documentation for the  python3 branch](https://github.com/stoqs/stoqs/blob/f700e8b6b52d99ceab16ceefac279975adde93ac/doc/instructions/PRODUCTION.md) (still a work in progress) that I would dearly love to replace with a Docker recipe. For testing this, port 8000 should be mapped to the host.

**Comment by carueda** [comment]
>  I think this can be a standard nginx image that provides a uWSGI "gateway" to the STOQS application. 

Because of all the provisioning required and already done on the base image, I'm first trying to do a regular `yum -y install nginx` and run nginx instead of apache. But we can certainly revisit the possibility of using the official nginx image later on.

Now, while trying to install nginx on the main image I'm getting the following error:

```
Transaction check error:
  file /var/log/nginx from install of nginx-1:1.10.2-1.el7.x86_64 conflicts with file from package InstantReality-RedHat-7-x64-2.8.0-38619.x86_64
```

I'll try to search for possible ways to avoid the conflict, but I'm wondering whether InstantReality could be skipped, at least during this initial exploration phase.

**Comment by carueda** [comment]
I went ahead with skipping InstantReality for now, and now the base image's entry point is the nginx server. Opening `http://localhost:8000` I can now see the default page, which btw provides useful info for next steps about incorporating relevant STOQS config files:

> This is the default index.html page that is distributed with nginx on Fedora. It is located in `/usr/share/nginx/html`.
>
>You should now put your content in a location of your choice and edit the root configuration directive in the nginx configuration file `/etc/nginx/nginx.conf`.

**Comment by MBARIMike** [comment]
InstantReality is needed only for preparing meshes of terrain data ([step 4 in SPATIAL_3D.md](https://github.com/stoqs/stoqs/blob/f700e8b6b52d99ceab16ceefac279975adde93ac/doc/instructions/SPATIAL_3D.md)) for presentation in the Spatial -> 3D window. It's (likely) not needed on a production installation.

**Comment by carueda** [comment]
Status: 

- Images:
  - mbari/stoqs  (based on mbari/stoqs-base)
  - mbari/stoqs-postgis
  - geodata/mapserver:7.0.1
  - rabbitmq:3

- STOQS django GUI running and exposed on host: http://localhost:8000/  (with empty Campaigns table at the moment) 

    ![image](https://user-images.githubusercontent.com/556505/31084355-0a6d151c-a749-11e7-9918-2e73d6dc9b5f.png)

Next:
- Mechanism to load data

**Comment by carueda** [comment]
@MBARIMike Hi Mike, the basic programs we tried earlier today are now running under the stoqs container, but with the "postgres" username (not "stoqsadm"):

```
       docker exec -it stoqs bash
 
        cd stoqs/
        ./manage.py makemigrations stoqs --settings=config.settings.ci --noinput
        ./manage.py migrate --settings=config.settings.ci --noinput --database=default
        wget -q -N -O loaders/Monterey25.grd http://stoqs.mbari.org/terrain/Monterey25.grd
        loaders/loadTestData.py
```

I captured the whole output of the above commands here:
https://gist.github.com/carueda/ece1696de65dd08b10432735e2158f9d

The GUI now looks like this: 
![image](https://user-images.githubusercontent.com/556505/31148689-57c82a82-a842-11e7-83b1-83b1f5c04284.png)

and after clicking the link there:

![image](https://user-images.githubusercontent.com/556505/31148817-b9f8ab6e-a842-11e7-9367-0143b1a19143.png)

So, great progress! ... I'll look next into the mapserver settings.

**Comment by carueda** [comment]
With some adjustments, those mapserver errors in the browser console above are fixed:

![image](https://user-images.githubusercontent.com/556505/31152003-f04fec0c-a84e-11e7-99da-d46fb050aa79.png)

**Comment by carueda** [comment]
Hi Mike,  thanks for stopping by again. So, one of the next steps is to make that /dev/shm location configurable, and re-enable the corresponding volume mapping in the stoqs service in docker-compose.yml.

**Comment by MBARIMike** [comment]
This looks great!

There's likely some things I need to change wrt to the MAPFILE_DIR so that the stoqs and mapserver containers share a common location for writing and reading (via a url path) the mapfiles. 

I comfortable merging https://github.com/stoqs/stoqs/pull/641 so that I can more easily build on this work.

**Comment by MBARIMike** [comment]
Just a quick note to indicate that I've picked up work on this issue with https://github.com/stoqs/stoqs/pull/671 and https://github.com/stoqs/stoqs/pull/672.

**Comment by MBARIMike** [comment]
I'm struggling with devising a secure way to give shell scripts and python programs the postgres superuser permission to create new databases. The issue seems to boil down to making all postgres connections via TCP/IP ('host' in the pg_hba.conf file) in a docker-compose deployment architecture. In the Vagrant VM installation postgres is available via 'local' entries in the pg_hba.conf file, providing more options to connect with the proper superuser permission to create new databases.

It seems to me that breaking up the STOQS application into separate containers is a good thing to do, but I need to understand the best (most secure) way to make database connections so that the database creation code will work in all environments, e.g. in [test.sh](https://github.com/stoqs/stoqs/blob/2bfee6ab584e14e0c10287c74035da35db3af47e/test.sh#L51-L55) and [stoqs/loaders/load.py](https://github.com/stoqs/stoqs/blob/2bfee6ab584e14e0c10287c74035da35db3af47e/stoqs/loaders/load.py#L62-L64)

**Comment by MBARIMike** [comment]
I seem to have a solution to use a /root/.pgpass file, added this to Dockerfile-stoqs:
```
# Allow for psql execution (used for database creation) without a password
RUN echo ${STOQS_PGHOST}:\*:\*:postgres:${POSTGRES_PASSWORD} > /root/.pgpass &&\
    chmod 600 /root/.pgpass
```
But this will put the POSTGRES_PASSWORD in the Docker image, which I don't want. @carueda, you mentioned that there is a way to have just the /root/.pgpass file put into a Docker volume. Can you point me in a direction that tells me how to do that?

**Comment by carueda** [comment]
@MBARIMike  yes, the idea would be to add an entry under the db volumes section in `docker-compose.yml` like this:

```yml
  postgis:
    volumes:
      - ${STOQS_VOLS_DIR}/pgdata:/var/lib/postgresql/data
      - ${PG_PASS_FILE}:/root/.pgpass
```

Where  `${PG_PASS_FILE}` (or whatever other env var name for this purpose) points to the location of the password file on the user's host machine, for example, `/$(pwd)/pg_password_file`.

As you already said, the key is to make this setting a runtime one (not something captured in the image anyway).

**Comment by MBARIMike** [comment]
Thanks! I think it's working better now with [this commit](https://github.com/stoqs/stoqs/pull/672/commits/818600b0a2bcaad5206722754b13609a404b13d1).

**Comment by MBARIMike** [comment]
A suggestion came up in our group meeting about doing Docker builds in my Vagrant VM vs. on the host OS. A stoqs image build is almost twice as fast on my MacOS: 6 minutes vs. 4 minutes.

**Comment by MBARIMike** [comment]
Woohoo! With https://github.com/stoqs/stoqs/pull/680 mapserver is now working!

**Comment by MBARIMike** [comment]
Found project showing how to integrate nginx: https://github.com/erroneousboat/docker-django

**Comment by MBARIMike** [comment]
This [PR](https://github.com/stoqs/stoqs/pull/687) adds a Production Docker installation to the [README](https://github.com/stoqs/stoqs/blob/master/README.md). There are no doubt still a few things to tune as I work to roll this out to production, but I think this issue is mostly solved. Thanks @carueda !

**Comment by MBARIMike** [comment]
Testing on a new VM with docker-compose installed (also works in the STOQS Vagrant VM):
```
git pull https://github.com/stoqs/stoqs.git
cd docker
docker-compose up
```
Open a browser on the VM and navigate to `https://localhost/default/query/` to see the default database (you will have to make a security exception as a generic certificate is used for the localhost connection).

In order to make the server work from a named server I'm envisioning that only changes will need to be made to settings in the [docker/.env](https://raw.githubusercontent.com/stoqs/stoqs/master/docker/.env) file. Starting now on documenting what those changes should be via comments in that file.

**Comment by MBARIMike** [comment]
With [this commit](https://github.com/stoqs/stoqs/pull/696/commits/fa0e5634dfb7426b8a56e1d46d330560c919f7f3) I get odd errors seemingly buried in the db_router.py code. More investigation is needed...

**Comment by MBARIMike** [comment]
Having `db_router.py` immediately return a HttpResponseBadRequest fixed the odd errors.

https://github.com/stoqs/stoqs/commit/b4120e82b9787c04f02e9d901ee7fadbac1c33ed#diff-3bde85ad6669b226430422fbcbabcc3c

**Comment by MBARIMike** [comment]
After getting everything to work from 4 Docker containers(postgis, mapserver, stoqs, & nginx) we are now adopting the suggestion made in the "Not containerizing nginx" section of http://www.eidel.io/2017/07/10/dockerizing-django-uwsgi-postgres/.

I suspect that there is an advantage for containerized nginx on a localhost-only STOQS. I need to think about the proper documentation to steer people toward not containerizing nginx on serious production servers...

**Comment by MBARIMike** [comment]
Comments have been added in docker/template.env about running nginx on the system and not in a container if certificate leakage is a concern.

**Comment by MBARIMike** [comment]
Doing final testing now of [these instructions](https://github.com/stoqs/stoqs#production-deployment-with-docker).
